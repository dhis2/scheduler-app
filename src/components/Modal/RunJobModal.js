import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { useDataMutation } from '@dhis2/app-runtime'
import {
    Button,
    Modal,
    ModalContent,
    ModalActions,
    ButtonStrip,
    NoticeBox,
} from '@dhis2/ui'
import i18n from '@dhis2/d2-i18n'

const RunJobModal = ({ id, hideModal, onComplete }) => {
    const [mutation] = useState({
        resource: `jobConfigurations/${id}/execute`,
        type: 'create',
    })
    const [runJob, { loading, error }] = useDataMutation(mutation, {
        onComplete: () => {
            hideModal()
            onComplete()
        },
    })

    return (
        <Modal open small onClose={hideModal}>
            <ModalContent>
                {error && (
                    <NoticeBox error title={i18n.t('Error running job')}>
                        {error.message}
                    </NoticeBox>
                )}
                <p>{i18n.t('Are you sure you want to run this job?')}</p>
            </ModalContent>
            <ModalActions>
                <ButtonStrip end>
                    <Button
                        name="hide-modal"
                        secondary
                        onClick={hideModal}
                        disabled={loading}
                    >
                        {i18n.t('Cancel')}
                    </Button>
                    <Button
                        name={`run-job-${id}`}
                        primary
                        onClick={runJob}
                        loading={loading}
                    >
                        {i18n.t('Run')}
                    </Button>
                </ButtonStrip>
            </ModalActions>
        </Modal>
    )
}

const { func, string } = PropTypes

RunJobModal.propTypes = {
    hideModal: func.isRequired,
    id: string.isRequired,
    onComplete: func.isRequired,
}

export default RunJobModal
