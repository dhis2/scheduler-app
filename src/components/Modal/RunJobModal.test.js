import React from 'react'
import { shallow, mount } from 'enzyme'
import { useDataMutation } from '@dhis2/app-runtime'
import RunJobModal from './RunJobModal'

jest.mock('@dhis2/app-runtime', () => ({
    useDataMutation: jest.fn(),
}))

afterEach(() => {
    jest.resetAllMocks()
})

describe('<RunJobModal>', () => {
    it('renders without errors', () => {
        useDataMutation.mockImplementation(() => [
            jest.fn(),
            { loading: false, error: null },
        ])

        const props = {
            id: 'id',
            hideModal: () => {},
            onComplete: () => {},
        }

        shallow(<RunJobModal {...props} />)
    })

    it('calls hideModal when cancel button is clicked', () => {
        useDataMutation.mockImplementation(() => [
            jest.fn(),
            { loading: false, error: null },
        ])

        const props = {
            id: 'id',
            hideModal: jest.fn(),
            onComplete: () => {},
        }
        const wrapper = mount(<RunJobModal {...props} />)

        wrapper.find('button').find({ name: 'hide-modal' }).simulate('click')

        expect(props.hideModal).toHaveBeenCalled()
    })

    it('calls hideModal when backdrop is clicked', () => {
        useDataMutation.mockImplementation(() => [
            jest.fn(),
            { loading: false, error: null },
        ])

        const props = {
            id: 'id',
            hideModal: jest.fn(),
            onComplete: () => {},
        }
        const wrapper = mount(<RunJobModal {...props} />)

        // Not a stable selector, but the backdrop does not have a data-test attribute
        wrapper.find('.backdrop').simulate('click')

        expect(props.hideModal).toHaveBeenCalled()
    })

    it('runs the expected tasks after a click on run job', async () => {
        const resolvedPromise = Promise.resolve()
        const onCompleteSpy = jest.fn()
        const hideModalSpy = jest.fn()
        const mutateSpy = jest.fn(() => resolvedPromise)
        let onComplete
        useDataMutation.mockImplementation((mutation, options) => {
            onComplete = options.onComplete
            return [mutateSpy, { loading: false, error: null }]
        })

        const props = {
            id: 'id',
            hideModal: hideModalSpy,
            onComplete: onCompleteSpy,
        }
        const wrapper = mount(<RunJobModal {...props} />)

        wrapper
            .find('button')
            .find({ name: 'run-job-id', type: 'button' })
            .simulate('click')

        await resolvedPromise
        expect(mutateSpy).toHaveBeenCalled()

        onComplete()
        expect(hideModalSpy).toHaveBeenCalled()
        expect(onCompleteSpy).toHaveBeenCalled()
    })
})
