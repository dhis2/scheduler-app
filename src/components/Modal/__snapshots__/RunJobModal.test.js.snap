// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RunJobModal> renders correctly 1`] = `
<Modal
  onClose={[Function]}
  open={true}
  small={true}
>
  <Content>
    Are you sure you want to run this job?
  </Content>
  <Actions>
    <ButtonStrip
      end={true}
    >
      <Button
        name="hide-modal"
        onClick={[Function]}
        secondary={true}
        type="button"
      >
        Cancel
      </Button>
      <Button
        name="run-job-id"
        onClick={[Function]}
        primary={true}
        type="button"
      >
        Run
      </Button>
    </ButtonStrip>
  </Actions>
</Modal>
`;
